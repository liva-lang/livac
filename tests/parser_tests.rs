use insta::assert_snapshot;
use livac::{lexer::tokenize, parser::parse};
use std::fs;

/// Test helper para casos correctos del parser
fn test_parser_ok(test_name: &str) {
    let source = fs::read_to_string(format!("tests/parser/ok_{}.liva", test_name))
        .unwrap_or_else(|_| panic!("Failed to read test file: ok_{}.liva", test_name));
    
    let tokens = tokenize(&source).unwrap();
    let ast = parse(tokens, &source).unwrap();
    
    // Convertir AST a JSON para snapshot
    let json = serde_json::to_string_pretty(&ast).unwrap();
    assert_snapshot!(format!("ok_{}.ast", test_name), json);
}

/// Test helper para casos de error del parser
fn test_parser_err(test_name: &str) {
    let source = fs::read_to_string(format!("tests/parser/err_{}.liva", test_name))
        .unwrap_or_else(|_| panic!("Failed to read test file: err_{}.liva", test_name));
    
    let tokens = tokenize(&source).unwrap();
    let result = parse(tokens, &source);
    assert!(result.is_err(), "Expected parser error for test: {}", test_name);
    
    let error_msg = result.unwrap_err().to_string();
    assert_snapshot!(format!("err_{}.diag", test_name), error_msg);
}

#[test]
fn test_functions_oneliner() {
    test_parser_ok("functions_oneliner");
}

#[test]
fn test_functions_block() {
    test_parser_ok("functions_block");
}

#[test]
fn test_classes() {
    test_parser_ok("classes");
}

#[test]
fn test_control_flow() {
    test_parser_ok("control_flow");
}

#[test]
fn test_expressions() {
    test_parser_ok("expressions");
}

#[test]
fn test_imports() {
    test_parser_ok("imports");
}

#[test]
fn test_unclosed_paren() {
    test_parser_err("unclosed_paren");
}

#[test]
fn test_unclosed_brace() {
    test_parser_err("unclosed_brace");
}

#[test]
fn test_case_without_switch() {
    test_parser_err("case_without_switch");
}

#[test]
fn test_return_outside_function() {
    test_parser_err("return_outside_function");
}

#[test]
fn test_duplicate_default() {
    test_parser_err("duplicate_default");
}
