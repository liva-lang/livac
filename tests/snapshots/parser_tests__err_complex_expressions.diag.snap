---
source: tests/parser_tests.rs
expression: error_msg
---
[1;31mโ[0m [1;31mE2000[0m: [1mParse Error[0m [90m[Parser][0m
[90mโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโ[0m
  [1;34mโ[0m [36m<input>[0m:[1;33m4[0m:[1;33m13[0m

  [90m   2[0m [90mโ[0m main() {
  [90m   3[0m [90mโ[0m     // Error: acceso a miembro sin objeto
  [90m   4[0m [90mโ[0m
  [90m    [0m [90mโ[0m let x = .name
  [90m    [0m [90mโ[0m         [1;31m^[0m
  [90m   5[0m [90mโ[0m 
  [90m   6[0m [90mโ[0m     // Error: llamada a funciรณn sin funciรณn
  [90m    [0m [90mโ[0m

  [1;34mโ[0m Expected expression

  [33m๐ก[0m [97mCheck for missing semicolons, parentheses, or keywords[0m

  [34m๐[0m Learn more: [36mhttps://liva-lang.org/docs/errors/parser#e2000[0m
[90mโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโโ[0m
